### ISSUE TITLE
Bugs related to invisibility

### ISSUE THREAD

**Issue by adaneslavo** [LoneGazebo/Community-Patch-DLL#10398]
VP 4.2.5:

I created a promotion and a counterpromotion giving the invisibility on demand.
![2023-10-26_22h13_43](https://github.com/LoneGazebo/Community-Patch-DLL/assets/33016300/4a23754d-058b-4855-a9fb-e9709a5fe17a)
![2023-10-26_22h13_29](https://github.com/LoneGazebo/Community-Patch-DLL/assets/33016300/d2e356c4-30dc-413d-99b4-4866435c7874)
Now I went to the city, and declared war to the city-state. Few weird behaviours were met then:
- enemy unit with counterpromotion didn't see the invisible units;
- cities do not attack invisible units (they do that only when my unit attacks the city);
- units which lose invisibility are still invisible to other units, but... when they were invisible enemy units couldn;t do anything, but when they lost it, enemy units entered their tile, pushed them to adjacent tile without harming and next turn still didn't attack, but went into the city;

So from what I can see, Invisibility while given, it cannot be removed, even when promotion is not longer available;
Invisibility different than base one for submarines is not recognized and thus not countered.

Some pictures for the better reception:
_Enemy Pikeman moved to the city in 2 turns (green arrow). My warrior was moved (blue arrow):_
![obraz](https://github.com/LoneGazebo/Community-Patch-DLL/assets/33016300/c421f6e4-1c9f-415d-856b-c04d711c25d9)

Enemy archer can only shoot the Artillery (it was not given the invisibility before). Warrior nor the Pikeman (they had the invisibility, but not anymore) cannot be attacked by the city nor the Archer:
![obraz](https://github.com/LoneGazebo/Community-Patch-DLL/assets/33016300/a52712aa-e922-46b6-ba0e-d9424ac95c22)

**Comment by n-core** [comment]
It's not a bug, it's more like a hardcoded limitation within the DLL.
You can't add invisible units besides Submarines (by using INVISIBLE_SUBMARINE).

Me and @azum4roll visited this stuff a while back, and it could be possible to make the InvisibleInfos table moddable.
You might want to investigate it yourself by looking at the cpp files itself.

**Comment by n-core** [comment]
This should be marked as Feature Request, not bugs.

**Comment by Tekamthi** [comment]
Yeah I've played around with this as well... My conclusion: invisible_submarine is the only one that is truly useable, unless we wanted some kind of invisible spy units etc.

Even if you use invisible_submarine, but have it turn on and off (ie same unit has promo that enables it, and then loses this promo), the game UI doesn't recognize this properly, and at least to human player the now-visible unit's 3d model will appear and disappear on the map as though it were still invisible.  I infer this is game engine stuff beyond the .DLL even.

One theoretical idea I have not tested... Unit tables have fields for IsConvertUnit, which appears to trigger off an hp threshold, ie unit drops below x hp, it transforms to another unit entirely.  Again I have never tested, but theorize you could use this to toggle invisibility on and off.  It's probably the only way without using Lua (though you could just do same in Lua much more flexibly)

**Comment by n-core** [comment]
The reason why you can't add more `InvisibleType` is because while `INVISIBILE_SUBMARINE` ID is 0, if you add another invisible type, then the ID is 1. Unfortunately, `NUM_INVISIBLE_TYPES` ID value is 1 in the cpp, so this is the hardcoded limitation in the cpp function.

We need to make it so that `NUM_INVISBLE_TYPES` doesn't conflict with its value when adding more invisible types.

**Comment by adaneslavo** [comment]
Another Spaghetti code by Firaxis.

**Comment by Hokath** [comment]
It turns out that
MOD_GLOBAL_EXTRA_INVISIBLE_TYPE will turn NUM_INVISIBLE_TYPES from 1 to... 2
However in my test the invisibility you get is like total invisibility -- being adjacent to the unit doesn't reveal it.

**Comment by Tekamthi** [comment]
testing various configs with invisibility again recently -- only using submarine version since thats the only one that works... i can toggle a unit from invisible and back to visible, with the unit showing up correctly in the UI in both states, using lua to assign everything, keeping the invis promo out of the free promos table, and upgrading the unit to itself whenever the unit is required to toggle back to visible.

This workaround might be obvious to some but it would be nice if at least this much could be handled within the .dll.  If nothing else we should be able to toggle it on/off, can be used in the pre-submarine eras as an additional effect

**Comment by Tekamthi** [comment]
revisiting this recently, it seems that invisibility type can never be reset to -1 --> normal unit starts with invis type = -1; if given invis promo it gets invis type = 0; if invis promo is removed, invis type = 0 still, it never changes back.  If this can be fixed, even just for the existing invis type, we're incrementally further ahead on this issue imo

**Comment by azum4roll** [comment]
There's more than a dozen effects that can't be removed from a unit for some reason.

<img width="1050" height="537" alt="Image" src="https://github.com/user-attachments/assets/f3778668-46f7-4eec-9333-d197ce87014d" />

**Comment by adaneslavo** [comment]
Why would we want to remove them?
Isnt that the promo defines if unit has the ability or not?

So here Invisibility is hardcoded?

**Comment by azum4roll** [comment]
There's logic defined when a promotion is removed.

**Comment by Tekamthi** [comment]
> Why would we want to remove them?

In the storms mod I was thinking of having a location promo toggle invis (even just the sub variety) onto units passing through the storm plots.  Effectively confirmed the logic posted above via trial and error: can change a unit from visible to invisible, but the only method to toggle back to visible is to recreate the unit.

I do this in one of my recon mods when the unit engages in combat; because recon are relatively few in number it works alright, just with some minor caveats with auto-explore and fortify/heal -- I may try that same method with storms eventually but its kinda a clunky workaround, especially to apply so broadly. Would be more elegant on the modmod side if the invis value could just toggle back -- but I infer there may be some deeper limitations

Could also maybe be used in some kind of special ability to make subs temporarily visible, or maybe a woodsman 2 that made infantry invisible only in forest/jungle etc.
